From e995fae1f750b1a12ef2941808d710dd58f17691 Mon Sep 17 00:00:00 2001
From: Amit Prakash Ambasta <amit.prakash.ambasta@gmail.com>
Date: Thu, 2 Jun 2022 11:08:17 +0530
Subject: [PATCH 1/2] www/firefox www/firefox-esr mail/thunderbird: fix build
 with libc++ 14

Mozilla attempts to redefine several internal libc++ functions, such as
__throw_out_of_range(). With libc++ 14 and higher, this now leads to
compile errors:

In file included from /wrkdirs/usr/ports/www/firefox/work/.build/dist/system_wrappers/deque:3:
/usr/include/c++/v1/deque:1865:9: error: call to '__throw_out_of_range' is ambiguous
        _VSTD::__throw_out_of_range("deque");
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__config:824:15: note: expanded from macro '_VSTD'
              ^
/wrkdirs/usr/ports/www/firefox/work/.build/dist/include/mozilla/throw_gcc.h:97:59: note: candidate function
MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_out_of_range(
                                                              ^
/usr/include/c++/v1/stdexcept:264:6: note: candidate function
void __throw_out_of_range(const char*__msg)
     ^

To work around this issue, avoid redefining those internal libc++
functions, if _LIBCPP_VERSION is set.

Signed-off-by: Amit Prakash Ambasta <amit.prakash.ambasta@gmail.com>
---
 memory/mozalloc/throw_gcc.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/memory/mozalloc/throw_gcc.h b/memory/mozalloc/throw_gcc.h
index aeeb770ba7..bc1f5b0a81 100644
--- a/memory/mozalloc/throw_gcc.h
+++ b/memory/mozalloc/throw_gcc.h
@@ -74,6 +74,7 @@ __throw_bad_function_call(void) {
   mozalloc_abort("fatal: STL threw bad_function_call");
 }
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_logic_error(
     const char* msg) {
   mozalloc_abort(msg);
@@ -98,12 +99,14 @@ MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_out_of_range(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif
 
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_runtime_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
 
+#if !defined (_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_range_error(
     const char* msg) {
   mozalloc_abort(msg);
@@ -118,6 +121,7 @@ MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void
 __throw_underflow_error(const char* msg) {
   mozalloc_abort(msg);
 }
+#endif
 
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_ios_failure(
     const char* msg) {
-- 
2.35.1

